src/sentry/static/sentry/app/components/modals/featureTourModal.tsx

```diff
@@ -0,0 +1,145 @@
+import React from 'react';
+import omit from 'lodash/omit';
+import styled from '@emotion/styled';
+
+import {openModal, ModalRenderProps} from 'app/actionCreators/modal';
+import Button from 'app/components/button';
+import ButtonBar from 'app/components/buttonBar';
+import {t} from 'app/locale';
+import {IconClose} from 'app/icons';
+import {callIfFunction} from 'app/utils/callIfFunction';
+import space from 'app/styles/space';
+
+type TourStep = {
+  title: string;
+  image?: React.ReactNode;
+  body: React.ReactNode;
+  actions: React.ReactNode;
+};
+
+type ContentsProps = ModalRenderProps & {
+  /**
+   * The list of tour steps.
+   * The FeatureTourModal will manage state on the active step.
+   */
+  steps: TourStep[];
+  /**
+   * Triggered when the tour is advanced.
+   */
+  onAdvance?: (currentIndex: number, durationOpen: number) => void;
+  /**
+   * Triggered when the tour is closed by completion or x
+   */
+  onCloseModal?: (currentIndex: number, durationOpen: number) => void;
+};
+
+type ContentsState = {
+  /**
+   * The current step offset to show
+   */
+  current: number;
+
+  /**
+   * The timestamp with ms the tour was opened,
+   * used to track duration on each step.
+   */
+  openedAt: number;
+};
+
+class ModalContents extends React.Component<ContentsProps, ContentsState> {
+  state: ContentsState = {
+    openedAt: Date.now(),
+    current: 0,
+  };
+
+  handleAdvance = () => {
+    const {onAdvance} = this.props;
+    this.setState({current: this.state.current + 1}, () => {
+      const duration = Date.now() - this.state.openedAt;
```

Comment:
shouldn't we perform this operation if the `componentDidUpdate`?

Reply:
I considered that but didn't want to have to write more logic to detect that the current state value had changed. Is triggering component callbacks from setState() callbacks problematic in other ways?